### ISSUE TITLE
Upgrading/shopping/drafting limits

### ISSUE THREAD

**Issue by tukkek** [tukkek/javelin#9]
From e-mail and comment https://javelinrl.wordpress.com/2016/04/30/javelin-1-4-bug-fix-release/comment-page-1/#comment-59

**Comment by tukkek** [comment]
> Little bug for you that seems to really limit the game: In terms of upgrading, the ability list runs out and then disapears once you research so many upgrades. This typically limits class selection to 1 class and only a handful of unevenly picked abilities. The ‘redraw all choices’ option in research does not generate a new list, so if theres say only 3 items left on the list of choices, and upgrades have already been removed, there is no way to get new upgrades. So the solution would be to make it so when you ‘redraw all choices’, the list refreshes entirely even if upgrades have been removed/used up. If you do fix this, remember that you’re going to have to deal with the fact that players will have their upgrade screen eventually flood with zillions of upgrades. (Might want to make it so they can voluntarily remove upgrades from there list to combat being overwhelmed)

At first there was no limit on how many upgrades or anything else you could have in a single Town but then I realized that if it was too much soon the game would run out of keys to assign to each of them. The game uses 1-9 then a-z for keys when you enter a selection screen (like the Town's upgrade and shopping screens).

There are two ways to circumvent this:

1. Limit the number of upgrades/shopping items/drafting options per Town (which is what I did, up to 20 of each for now).

2. Create a pagination system so that the user could "go to the next page" and not be limited to only 20 or so options. This is what I wanted to do on version 2.0, which would enable the player to use the mouse as well, so we wouldn't need to have 1 key per upgrade/item/monster.

There is another question though: do we really want the player to be able to have all the upgrades and items on a single Town? Should there be a limit? I think it's nice to have a limit because it means that the player will need to conquer other Towns in order to have more upgrades and items to take advantage of. This forces the player to explore and conquer the map, which is one of the goals/challenges of the game.

If the player can stay on one town the entire time it will be easier for him to never move out and explore the game world, which would make Javelin a less interesting game. But this is just my opinion, let me know what you think - the game is certainly open to new ideas!

For example: another idea i had was to let the player create new towns on the map, which would enable him to have more towns and more upgrades to choose from (and it would make the game feel a bit more like Civilization, which is also a good thing).

Another idea: instead of having a fixed limit (20) we could have a maximum of 10 + town size (or 5 + town size). So a starting town could only have 11 (or 6) upgrades/shop items but as the town grows it would be able to offer more options accordingly.

**Comment by tukkek** [comment]
Also, since you mentioned being interested in developing the game and helping with QA these hacks should help you out a lot in this regard. Just put them in your `preferences.properties` file (one per line) in order to help you test out the game:

```
cheat.gold=9000
cheat.xp=500
cheat.labor=9000 #labor for town research
cheat.haxor=9000 #tickets for The Temple of Haxor
cheat.combat=false #no random encounters + enemies don't move on the map
cheat.map=show #shows the entire world map
```

To disable any of these just put a # on the start of the respective line. For example to disable all of them you would have:

```
#cheat.gold=9000
#cheat.xp=500
#cheat.labor=9000 #labor for town research
#cheat.haxor=9000 #tickets for The Temple of Haxor
#cheat.combat=false #no random encounters + enemies don't move on the map
#cheat.map=show #shows the entire world map
```

If you're using Windows please use Wordpad, Microsoft Office Word or Notepad++ to edit the preferences file. Just using normal Notepad (which comes with Windows) will probably make a mess out of your file.

**Comment by Zireael07** [comment]
I would like to help with QA too. Do those work in 1.6?

**Comment by tukkek** [comment]
Hi Zireael! Your fame precedes you! I've tried playing your game (veins of the earth) a couple of times already but being on Linux and not having much patience with Wine I actually never succeeded in doing so. I knew of your game already and was very proud to have you as one of the first people starring Javelin here on GH :)

If you want a more up-to-date list of the development options please check the wiki https://github.com/tukkek/javelin/wiki/Development-options

It would be a great help to have you on-board! I'll be releasing an alpha for 1.7 with basic mouse support (already working) and a new location which another tester is waiting to playtest and help design! I suggest you wait until the end of the week so I can release it properly (it was supposed to be ready already but a nasty flu delayed my plans a bit)...

If you want me to send you a ping once the 1.7 alpha is ready send an email to javelinrl@gmail.com - that would be easier but of course I can let you know over here too.

Also if you want more info on the new location which is already half-way done see #21

**Comment by tukkek** [comment]
The entire city system has been overhauled for 1.7, making this specific issue entirely invalid. However, the new district system is much better designed and not only fixes the issues of too many option per screen by offering more locations, but also solves the issue of not allowing for many upgrades by greatly increasing the number of academies/shops/etc available in a single town, while also putting a lot of that on the hands of the player, if he wants to take control over the process.
